rbd: copy passphrase for encrypted clones

nixpanic opened the pull request:
When a source volume is encrypted, the passphrase needs to be copied and
stored for the newly cloned volume.

This requires the `rbdImage.encryption` component to be configured, and that
is what most of the changes in this PR make possible.

Fixes: #1470

## Notes for reviewers

This PR contains a large number of commits. The easiest to review these is by going to https://github.com/ceph/ceph-csi/pull/1918/commits and clicking the title of the 1st change. Then use the `[next]` button in the top blue bar where the subject of the change is written to go to the next commit.

Each single commit is relatively simple, reviewing all changes at once will be close to impossible.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
rbd: copy passphrase for encrypted clones

When a source volume is encrypted, the passphrase needs to be copied and
stored for the newly cloned volume.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: add rbdImage.copyEncryptionConfig() to copy encryption metadata

Cloning volumes requires copying the DEK from the source to the newly
cloned volume. Introduce copyEncryptionConfig() as a helper for that.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: add VolumeEncryption.StoreCryptoPassphrase()

The new StoreCryptoPassphrase() method makes it possible to store an
unencrypted passphrase newly encrypted in the DEKStore.

Cloning volumes will use this, as the passphrase from the original
volume will need to get copied as part of the metadata for the volume.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: add validation for cloning encrypted volumes

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: pass secrets when creating an source rbdVolume for cloning

Without this, the rbdVolume can not connect to the Ceph cluster and
configure the (optional) encryption.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: improve logging in deleteImage()

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: move deletion of DEK to deleteImage()

The ControllerServer should not need to care about support for
encryption, ideally it is transparantly handled by the rbdVolume type
and its internal API.

Deleting the DEK was one of the last remainders that was explicitly done
inside the ControllerServer.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: log available configs when KMS not found

When the KMS configuration can not be found, it is useful to know what
configurations are available. This aids troubleshooting when typos in
the KMS ID are made.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: copy encryption configuration in CreateSnapshot

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: configureEncryption() in genSnapFromSnapID()

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: set kmsID in reserveSnap()

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: copyEncryptionConfig in doSnapshotClone()

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
logging: correct formatting when reporting error in createVolumeFromSnapshot()

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
rbd: move setupEncryption() from buildCreateVolumeResponse to CreateVolume

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
util: add logging when OpenEncryptedVolume() encounters an error

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
cleanup: release resources for rbdImages objects after use

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
/retest ci/centos/upgrade-tests-cephfs

nixpanic commented:
> /retest ci/centos/upgrade-tests-cephfs

Deploying the snapshot controller failed:
```
snapshotter controller creation failed
```

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -749,6 +749,36 @@ var _ = Describe("RBD", func() {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+
+				validatePVCClone(pvcPath, appPath, pvcSmartClonePath, appSmartClonePath, f)
```
can we validate clone is also encrypted to make sure the app gets the encrypted device when using clone PVC?

nixpanic replied:
I guess so, will see if I can add that when reworking the e2e commit

nixpanic replied:
this has been added now too

Madhu-1 commented on internal/journal/voljournal.go:
```diff
@@ -691,9 +691,22 @@ func (conn *Connection) GetImageAttributes(ctx context.Context, pool, objectUUID
-func (conn *Connection) StoreImageID(ctx context.Context, pool, reservedUUID, imageID string) error {
-	err := setOMapKeys(ctx, conn, pool, conn.config.namespace, conn.config.cephUUIDDirectoryPrefix+reservedUUID,
-		map[string]string{conn.config.csiImageIDKey: imageID})
+func (conn *Connection) StoreImageID(ctx context.Context, pool, reservedUUID, imageID, kmsID, owner string) error {
```
the function is doing more work than storing the imageID. can we have a separate function to store the encryption ID and owner ?

nixpanic replied:
Maybe this change is not needed, as `ReserveName()` should set the correct metadata already. I'll try dropping this change and see how it goes.

Madhu-1 replied:
Yes, correct. This is for post metadata reservation, Once the Reservation is done, we need to reserve the imageID at last after creating the image.

nixpanic replied:
Dropping this does not seem to affect the functioning, so that looks good :+1: This change will not be included in the next update.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -274,7 +271,16 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		}
-		return buildCreateVolumeResponse(ctx, req, rbdVol)
+
+		if parentVol != nil && parentVol.isEncrypted() {
```
the same need to be handled for snapshot also?

nixpanic replied:
no, that does not seem to be required

Madhu-1 replied:
what if during creating a clone from snapshot operation the rbd image got created but the encryption information is not stored (maybe in midway the controller restarted?)

nixpanic replied:
A next call would return a `VolumeOperationAlreadyExists` error, just like in other places where the provisioning is partially done and aborted. I do not think it is easy to recover from that automatically.

Madhu-1 replied:
I still feel like this is also needed for a snapshot. during vol Exist check for restoring rbd image from snapshot the parentVol is empty we need to check  if the volume encrypted if is restored from a snapshot

Madhu-1 replied:
```go
// check after vol exists from snapshot
if rbdSnap.isEncrypted() {
			rbdSnap.conn = rbdVol.conn.Copy()
			err = rbdSnap.copyEncryptionConfig(&rbdVol.rbdImage)
			if err != nil {
				return nil, fmt.Errorf("failed to clone encryption config: %w", err)
			}
		}
```

nixpanic replied:
I think this is now addressed with the added `repairEncryptionConfig()` function.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1071,6 +1106,29 @@ func (rv *rbdVolume) cloneRbdImageFromSnapshot(ctx context.Context, pSnapOpts *r
+	deleteClone := true
+	defer func() {
+		if deleteClone {
+			err = librbd.RemoveImage(rv.ioctx, rv.RbdImageName)
```
use rv.DeleteImage which uses manager tasks to delete the image?

nixpanic replied:
Not sure that is needed, it is only for an error case that the just created image needs to be removed again.

Because the `deleteImage()` helper uses CephMgr, it also needs access to the Credentials, which are not available here. I think it would clutter the code more than is useful.

Madhu-1 replied:
RemoveImage directly from the ceph is time-consuming that's why we brought a manager to handle the image deletion asynchronous. I think this is fine for now.

Madhu-1 commented on e2e/rbd.go:
```diff
@@ -749,6 +749,36 @@ var _ = Describe("RBD", func() {
+				if err != nil {
+					e2elog.Failf("failed to create storageclass with error %v", err)
+				}
+
```
Add an e2e to validate encrypted PVC+ snapshot +restore also

nixpanic replied:
oh, yes, I was planning to add that, but seem to have forgotton

nixpanic replied:
Done!

nixpanic committed:
cleanup: address "nolint" comments for RBD CreateSnapshot

Introduce helper function cloneFromSnapshot() that takes care of the
procedures that are needed when an existing snapshot has been found.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -226,6 +221,7 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
 }
 
 // CreateVolume creates the volume in backend.
+// nolint:gocyclo // encrypted cloning/snapshots added more complexity, this needs to be addressed
```
fwiw, I have an extra commit that cleans up the `nolint` annotation. It will be posted with the next update of this PR.

nixpanic committed:
cleanup: use buildCreateVolumeResponse() to simplify CreateVolume()

buildCreateVolumeResponse() exists exactly for the need to create a
csi.CreateVolumeResponse based on an rbdVolume. Calling this helper
reduces the code duplication in CreateVolume().

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
cleanup: move copyEncryptionConfig() from CreateVolume to Exists()

The rbdVolume that needs its encryption configured is constructed in the
Exists() method. It is suitable to move the copyEncryptionConfig() call
there as well, so that the object is completely constructed in a single
place.

Golang-ci:gocyclo complained about the increased complexity of the
Exists() function. Moving the repairing of the ImageID into its own
helper function makes the code a little easier to understand.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: add verification for encrypted Snapshot/Restore operations

This moves validatePVCSnapshot() into its own function, so that it
follows the same format as validatePVCClone() does already.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Madhu-1 commented:
@nixpanic E2E is failing

nixpanic commented:
> @nixpanic E2E is failing

```
Apr  8 18:10:15.871: FAIL: failed to create storageclass with error volumesnapshotclasses.snapshot.storage.k8s.io "csi-rbdplugin-snapclass" already exists
```

The Snapshot tests should also delete the snapshotclass once they are done. Will address it right away!

humblec commented on e2e/rbd_helper.go:
```diff
@@ -279,6 +268,37 @@ func validateEncryptedPVCAndAppBinding(pvcPath, appPath, kms string, f *framewor
+	return validateEncryptedImage(f, rbdImageSpec, app)
+}
+
+func validateEncryptedImage(f *framework.Framework, rbdImageSpec string, app *v1.Pod) error {
```
@nixpanic please add some source code comment about this function.

nixpanic replied:
will do!

humblec commented on internal/util/util.go:
```diff
@@ -319,3 +319,15 @@ func contains(s []string, key string) bool {
 	return false
 }
+
+func getKeys(m map[string]interface{}) []string {
```
please add some source code comment.

nixpanic replied:
ok!

humblec commented on internal/rbd/clone.go:
```diff
@@ -46,7 +46,10 @@ import (
 	// generate temp cloned volume
 	tempClone := rv.generateTempClone()
+	defer tempClone.Destroy()
+
```
nit: may be this extra line is not required

nixpanic replied:
not required, but with the extra line it makes it easier to read

humblec commented on internal/rbd/controllerserver.go:
```diff
@@ -274,7 +270,8 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 			}
 		}
-		return buildCreateVolumeResponse(ctx, req, rbdVol)
+
```
@nixpanic is this new line intentionally kept?

nixpanic replied:
Yes. After a closing `}`, if the next statement is not related to the previous code-block, I prefer an empty line to split the parts of the code.

humblec replied:
Ok.. Sure..

humblec commented on internal/rbd/encryption.go:
```diff
@@ -111,6 +111,41 @@ func (ri *rbdImage) setupEncryption(ctx context.Context) error {
 	return nil
 }
 
+func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
```
good to have a source code comment about this newly introduced functiion @nixpanic

nixpanic replied:
Sure, I'll add that.

humblec commented on internal/rbd/encryption.go:
```diff
@@ -111,6 +111,41 @@ func (ri *rbdImage) setupEncryption(ctx context.Context) error {
 }
 
+func (ri *rbdImage) copyEncryptionConfig(cp *rbdImage) error {
+	// get the unencrypted passphrase
```
Isnt this we are getting `crypt` passphrase . Got confused by above comment saying `unencrypted` passphrase

nixpanic replied:
yes, `unencrypted` is correct. Passphrases can be encrypted per volume, so a copy of the passphrase for the next volume will get (potentially) encrypted differently. The unencrypted passphrase will remain the same. I hope the comments inside the function make this process clear.

A new comment describing the function will also clarify this more.

nixpanic committed:
rbd: repair encryption config in case it is missing

It is possible that when a provisioner restarts after a snapshot was
cloned, but before the newly restored image had its encryption metadata
set, the new image is not marked as encrypted. This will prevent
attaching/mounting the image, as the encryption key will not be fetched,
or is not available in the DEKStore.

By actively repairing the encryption configuration when needed, this
problem should be addressed.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
e2e: only test a single encrypted clone/snapshot

The default number for cloning and snapshot/restore is 10 volumes. This
adds to the time the test suite runs. There is no need to validate 10
copies of the encrypted volume, a single copy is sufficient.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Madhu-1 commented on internal/util/crypto.go:
```diff
@@ -243,7 +249,10 @@ func EncryptVolume(ctx context.Context, devicePath, passphrase string) error {
-	_, _, err := LuksOpen(devicePath, mapperFile, passphrase)
+	_, stderr, err := LuksOpen(devicePath, mapperFile, passphrase)
+	if err != nil {
+		WarningLog(ctx, "failed to open LUKS device %q: %s", devicePath, stderr)
```
this should be an error log not a warning log? logging err with stderror might also be helpfull

nixpanic replied:
I'd leave that up to the caller. The caller can log the returned error with a level it deems appropriate. Only because it is a failure, I didn't think Info-level would be suitable.

Madhu-1 reviewed (dismissed):
LGTM

humblec commented:
@nixpanic mostly looks good! , one quick check here  would be updating the encryption doc is needed or not, I feel we can add some notes in the encryption doc about the encrypted clones. wdyt?

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

humblec commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

nixpanic committed:
logging: report issues in rbdImage.DEKStore API with stacks

It helps to get a stack trace when debugging issues. Certain things are
considered bugs in the code (like missing attributes in a struct), and
might cause a panic in certain occasions.

In this case, a missing string will not panic, but the behaviour will
also not be correct (DEKs getting encrypted, but unable to decrypt).
Clearly logging this as a BUG is probably better than calling panic().

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented on internal/rbd/snapshot.go:
```diff
@@ -89,7 +89,16 @@ func generateVolFromSnap(rbdSnap *rbdSnapshot) *rbdVolume {
+
+	if rbdSnap.isEncrypted() {
+		vol.conn = rbdSnap.conn.Copy()
+		err := rbdSnap.copyEncryptionConfig(&vol.rbdImage)
```
This is wrong for the case where the KMS uses the VolID for encrypting.decrypting the passphrase (SecretMetadataKMS). The encryption of the passphrase for the new rbdVolume is done with an incorrect VolID.

nixpanic commented:
Now most features manually tested with VaultTokenKMS and SecretMetadataKMS. The 2nd failed in CI, with issues that could be reproduced.

@Madhu-1 @humblec please review again (hopefully last time!)

nixpanic commented:
/retest ci/centos/mini-e2e

nixpanic commented:
> /retest ci/centos/mini-e2e

Deploying snapshot controller failed:
```
snapshotter controller creation failed
```

nixpanic commented:
/retest ci/centos/upgrade-tests-cephfs

nixpanic commented:
> /retest ci/centos/upgrade-tests-cephfs

Deploying snapshot controller failed:
```
snapshotter controller creation failed
```

nixpanic commented:
**ci/centos/mini-e2e-helm/k8s-1.20** failed with:
```
Apr  9 17:55:12.145: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
```

See #1969 

And so did **ci/centos/mini-e2e/k8s-1.19** https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.19/detail/mini-e2e_k8s-1.19/1098/pipeline

nixpanic commented:
Some e2e tests fail in **STEP: create ROX PVC clone and mount it to multiple pods**. This might be related to the cleanup of the VolumeSnapshotClass that was previously not done.

nixpanic commented:
> Some e2e tests fail in **STEP: create ROX PVC clone and mount it to multiple pods**. This might be related to the cleanup of the VolumeSnapshotClass that was previously not done.

All ready to go! Create/Delete the SnapshotClass when needed solved the failures.
